src/backend/distributed/executor/multi_executor.c

```diff
@@ -175,14 +175,53 @@ RouterCreateScan(CustomScan *scan)
 	else
 	{
 		Assert(isModificationQuery);
-		scanState->customScanState.methods = &RouterMultiModifyCustomExecMethods;
+
+		if (IsMultiRowInsert(workerJob->jobQuery))
+		{
+			/* multi-row INSERT is executed sequentially */
+			scanState->customScanState.methods = &RouterSequentialModifyCustomExecMethods;
+		}
+		else
+		{
+			scanState->customScanState.methods = &RouterMultiModifyCustomExecMethods;
+		}
 	}
 
 	return (Node *) scanState;
 }
 
 
 /*
+ * IsMultiRowInsert returns whether the given query is a multi-row INSERT.
+ */
+static bool
+IsMultiRowInsert(Query *query)
+{
+	ListCell *rteCell = NULL;
+	bool hasValuesRTE = false;
+
+	CmdType commandType = query->commandType;
+	if (commandType != CMD_INSERT)
+	{
+		return false;
+	}
+
+	foreach(rteCell, query->rtable)
+	{
+		RangeTblEntry *rte = (RangeTblEntry *) lfirst(rteCell);
+
+		if (rte->rtekind == RTE_VALUES)
```

Comment:
This probably needs explanatory comment: This is an accurate check only because `transformInsertStmt()` optimizes away single-row `VALUES()` in an `INSERT`. That's not necessarily obvious imo.